logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -792,8 +807,13 @@ private void releaseLockAndSwallow() {
     /**
      * return the {@link Page} for the next read operation.
      * @return {@link Page} will be either a read-only tail page or the head page.
+     * @throws QueueRuntimeException if queue is closed
      */
     public Page nextReadPage() {
+        if (isClosed()) {
+            throw new QueueRuntimeException(QueueExceptionMessages.WHILE_READING_NEXT_PAGE);
+        }
+
         lock.lock();
         try {
```

Comment:
For safety, we should do the check _after_ acquiring the lock else our check could be invalid by the time the lock has been acquired (yes, we happen to know that our `lock` of type `Lock` is a `ReentrantLock`, and the current callers of this public `nextReadPage()` method only do so after acquiring the reentrant lock, but these details aren't enough to make this public method safe on its own). 

In the write-case we _also_ check before acquiring the lock in order to "fail fast" and skip potentially-expensive serialization overhead, but here in the read-case I don't think that is necessary.

```suggestion
        lock.lock();
        try {
            if (isClosed()) {
                throw new QueueRuntimeException(QueueExceptionMessages.WHILE_READING_NEXT_PAGE);
            }

```

Reply:
Related: I see nothing outside of `Queue` calling `Queue#nextReadPage()`, and its details are tightly-coupled with the `Queue` implementation itself. Can it be made `private`?

Reply:
Yeah, how do you think if we make `Queue#nextReadPage` private and place queue-close check in `Queue#readPageBatch` after acquiring lock? Making queue close-check visible in reading batch operations would be more meaningful I think.

Reply:
This `Queue#nextReadPage` cannot reliably provide a "next read page" when the queue is closed so either it should validate that the queue is not closed or it should not be called when the queue is closed. Pushing the check down to `Queue#readPageBatch` requires sending it a `Page` that may or may not be meaningless.

`Queue#readPageBatch` _does_ check to see if the queue is closed, but the intent is different; it _assumes_ that the queue was open when it began reading, and only checks to see if the queue has _become_ closed after awaiting a fully-read head page to determine if it needs to stop trying to read and emit the already-read events.

Separately, the Queue really _should not_ return a `Page` to a thread that does not have exclusive access to it, which is what we allow by having a public method that acquires and releases the lock before returning a `Page`. This is the _only_ public method on `Queue` that exposes `Page` directly, and there is no reason for doing so.

Together, I think that:

 - `Queue#nextReadPage` should be `private` (to close abstraction-leak of `Page`, which is otherwise not a part of the public interface of `Queue`)
 - `Queue#nextReadPage` and should _assert_ that the lock is held by the current thread (with `ReentrantLock#isHeldByCurrentThread`, which would require the `Queue.lock` field type to be more specific) before calculating instead of acquire-calculate-release (to prevent the `Page` from being given to a caller that does not have exclusive access)
 - Either
   - a single queue-closed check should be done in `Queue#nextReadPage` after validating that the current thread has the lock (I favor this one because it is more DRY); or
   - a pair of queue-closed checks should be done in `Queue#readBatch` and `Queue#nonBlockReadBatch` after each acquires the lock.